rbd/mirror_group.go

```diff
@@ -0,0 +1,488 @@
+		cephIoctx(groupIoctx),
+		cGroupName,
+		C.bool(force))
+
+	return getError(ret)
+}
+
+// MirrorGroupPromote will promote the mirrored group to primary state
+//
+// Implements:
+//
+//	int rbd_mirror_group_promote(rados_ioctx_t p, const char *name,
+//	  							 uint32_t flags, bool force)
+func MirrorGroupPromote(groupIoctx *rados.IOContext, groupName string, force bool) error {
+	cGroupName := C.CString(groupName)
+	defer C.free(unsafe.Pointer(cGroupName))
+
+	rbdMirrorGroupPromote, rbdMirrorGroupPromoteErr := dlsym.LookupSymbol("rbd_mirror_group_promote")
+	if rbdMirrorGroupPromoteErr != nil {
+		return fmt.Errorf("%w: %w", ErrNotImplemented, rbdMirrorGroupPromoteErr)
+	}
+
+	ret := C.rbd_mirror_group_promote_dlsym(
+		rbdMirrorGroupPromote,
+		cephIoctx(groupIoctx),
+		cGroupName,
+		RBD_SNAP_CREATE_SKIP_QUIESCE,
+		C.bool(force))
+
+	return getError(ret)
+}
+
+// MirrorGroupDemote will demote the mirrored group to secondary state
+//
+// Implements:
+//
+//	int rbd_mirror_group_demote(rados_ioctx_t p, const char *name,
+//	  							uint32_t flags)
+func MirrorGroupDemote(groupIoctx *rados.IOContext, groupName string) error {
+	cGroupName := C.CString(groupName)
+	defer C.free(unsafe.Pointer(cGroupName))
+
+	rbdMirrorGroupDemote, rbdMirrorGroupDemoteErr := dlsym.LookupSymbol("rbd_mirror_group_demote")
+	if rbdMirrorGroupDemoteErr != nil {
+		return fmt.Errorf("%w: %w", ErrNotImplemented, rbdMirrorGroupDemoteErr)
+	}
+
+	ret := C.rbd_mirror_group_demote_dlsym(
+		rbdMirrorGroupDemote,
+		cephIoctx(groupIoctx),
+		cGroupName,
+		RBD_SNAP_CREATE_SKIP_QUIESCE)
+	return getError(ret)
+}
+
+// MirrorGroupResync is used to manually resolve split-brain status by triggering
+// resynchronization
+//
+// Implements:
+//
+//	int rbd_mirror_group_resync(rados_ioctx_t p, const char *name)
+func MirrorGroupResync(groupIoctx *rados.IOContext, groupName string) error {
+	cGroupName := C.CString(groupName)
+	defer C.free(unsafe.Pointer(cGroupName))
+
+	rbdMirrorGroupResync, rbdMirrorGroupResyncErr := dlsym.LookupSymbol("rbd_mirror_group_resync")
+	if rbdMirrorGroupResyncErr != nil {
+		return fmt.Errorf("%w: %w", ErrNotImplemented, rbdMirrorGroupResyncErr)
+	}
+
+	ret := C.rbd_mirror_group_resync_dlsym(
+		rbdMirrorGroupResync,
+		cephIoctx(groupIoctx),
+		cGroupName)
+	return getError(ret)
+}
+
+// MirrorGroupState represents the current state of the mirrored group
+type MirrorGroupState C.rbd_mirror_group_state_t
+
+// String representation of MirrorGroupState.
+func (mgs MirrorGroupState) String() string {
+	switch mgs {
+	case MirrorGroupEnabled:
+		return "enabled"
+	case MirrorGroupDisabled:
+		return "disabled"
+	case MirrorGroupEnabling:
+		return "enabling"
+	case MirrorGroupDisabling:
+		return "disabling"
+	default:
+		return "<unknown>"
+	}
+}
+
+const (
+	// MirrorGroupDisabling is the representation of
+	// RBD_MIRROR_GROUP_DISABLING from librbd.
+	MirrorGroupDisabling = MirrorGroupState(C.RBD_MIRROR_GROUP_DISABLING)
```

Comment:
I assume these are only included in very new `librbd.h` versions? Why not take the safe way as you have done for `RBD_SNAP_CREATE_SKIP_QUIESCE`?

Reply:
Done